### DISCUSSION TITLE
what is the best way to use a react component that makes xhr requests in multiple pages?

### DISCUSSION THREAD

**Discussion by vasanthi-d** [vercel/next.js#37027] [Help]
Hi All,

I just started with Next.js 12.x couple of weeks back and would like to get some input on how to best implement below scenario.

In my project, I have a react component. Let's say CompZ.  It has a prop 'searchText' of type String. Inside this component,
it makes multiple XHR calls in parallel. Calls looks like below where PARAM_1, PARAM_2 and PARAM_3 are constants (do not change) and {searchTex} is the only dynamic parameter.
 xhr1 = <<endpoint>>?query={searchText}&param=PARAM_1
xhr2 = <<endpoint>>?query={searchText}&param=PARAM_2
xhr3 = <<endpoint>>?query={searchText}&param=PARAM_3

Also, we have PageA, PageB, PageC

When CompZ is used in PageA, {searchText} will always be  'pageA'. 
When CompZ is used in PageB, {searchText} will always be  'pageB'. 
However, when CompZ is used in PageC, {searchText} can change based on user input.

I would like to know 
a) if it is possible to use getStaticProps in PageA and PageB so that these pages can be build during build time because parameters to the API are known ahead of time?

b) if this API provides server-side pagination feature, can we use getStaticProps anyway?

Any help is greatly appreciated.  Thanks.

**Comment by icyJoseph** [comment]
First I'd try to make the component generic, then if necessary, extract the core of the component into a custom hook, and use that custom hook wherever I need it. For example, considering only happy paths (no error handling):

```tsx
/* this is the important bit, this function is browser/server agnostic
    next.js polyfills fetch on the server! */
const fetchData = (query:string, param: string) => 
          fetch(`https://example.com/api?query=${query}&param=${param}`).then(res => res.json());
/* fetchData can be used in API routes, 
    GetStaticProps, GetServerSideProps, and even on the client (careful with CORS) */

/* so now we use it on a generic component
    assuming the data is a Record<string, string> or `null` by default */
type SearchWithParamsProps = {query: string, param: string, initial: null | Record<string, string>};
const SearchWithParams = ({ query, param, initial  = null}: SearchWithParamsProps ) => {
  // from here
  const [state, setState] = useState<Record<string, string> | null>(initial);

  useEffect(() => {
    let cancel = false;
    fetchData(query, param).then((data) => {
          if(cancel)  return; // don't set state if it was cancelled
          setState(data);
     });
    return () => {
      /* Important to cancel the request if query/param change - 
         read into Abort Controllers, or 
         use a cancel flag 
      */
     cancel = true;
   }
  }, [query, param]);

  // to here, could be a custom hook that returns state

  if (!state) return <div>Empty</div>

  return <div>Do something with state</div>
}

/* Now inside /pages/pageA.tsx or /pages/pageB.tsx */
const PageA = ({data}) => {
 // props contains the result of fetchData
 return <SearchWithParams query='pageA' param={PARAM_1} initial={data} />
}

export const getStaticProps = async () => {
   const data = await fetchData("pageA", PARAM_1);
   return { props: {data}}
}

/* For page C */
const PageC = () => {
 const [input, setInput] = useState("");

 return (
  <>
    <input value={value} onChange={e => setInput(e.target.value)}/> 
    <SearchWithParams query={input} param={PARAM_3} />
  </>
  );
}
```

> a) if it is possible to use getStaticProps in PageA and PageB so that these pages can be build during build time because parameters to the API are known ahead of time?

Yes. However, GetStaticProps, runs once at build time, and that is. If the data changes, you might want to revalidate that on the client, though this has SEO implications. It is often better to use the revalidate property on the return of `getStaticProps` so that the page is regenerated at runtime. More https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration.

> b) if this API provides server-side pagination feature, can we use getStaticProps anyway?

Yes. To fetch the most meaningful (usually first) page. How you navigate dictates what happens next... If each page of server data should have its own URL, it gets a bit more tricky,  not impossible! I assume you'd navigate through the pages of server data, on one client side page, and use queries and such to move the cursor, then you can use something similar to the above, querying a different page from the client. 

## GetStaticProps

GetStaticProps, for data that rarely changes, is awesome! The thing is that, the HTML generated with that data is what the client sees first, and what most Bots parse. So if the data has changed, even if you refresh it on the client, the HTML generated at build time, remains unchanged. That's where ISR kicks in, it allows the server to regenerate that HTML with fresh data by calling GetStaticProps at runtime, read more about it in the docs.

## SWR and React-Query

Reaching this point, I'd say that it might be better to learn and `SWR`, or `React-Query`, because these do a lot of the boilerplate above for you. See here https://swr.vercel.app/docs/with-nextjs#pre-rendering-with-default-data, for more information. These libraries also have pagination APIs and a bunch of other goodies that make it way easier to work with server data.

**Reply by vasanthi-d** [reply]
@icyJoseph Thank you very much your detailed response. It makes sense to me. Just to make sure I understood correctly, can you please clarify below?
1.  You suggest me to define a custom hook and have 'fetchData' with in the hook so it is accessible from API routes, getStaticProps and getServerSideProps. Correct? 

2. PageC will be a regular 'react' component without any getStaticProps or getServerSideProps. Correct?

3. When we have pagination, it will introduce another variable parameter like 'pageNumber' to the API.  We can determine value of this only based on user actions on UI.  So, in this case, is it still possible to have getStaticProps on PageA and PageB?

Truly appreciate your response. Thank you.

**Reply by icyJoseph** [reply]
Before answering, I'd say that it is important to try this out yourself first, now that you have some idea of where to start. Spin up a next.js project, which takes about 1 second, and try out the things you will do in your bigger project. 

Often in these discussions, people ask only about the things they consider to matter most, and leave out details that actually matter. Those who answer here can only go so far, and even if more detail was provided, sometimes we cannot afford the time to answer. So yeah, try this out, test your assumptions and mental model of Next.js in combination with React.

## Answers

> 1. You suggest me to define a custom hook and have 'fetchData' with in the hook so it is accessible from API routes, getStaticProps and getServerSideProps. Correct?

The custom hook is a React abstraction, to be used only within React function components.

The key thing here is `fetchData`. We could say that this is the smallest building block on my example. It can be used directly in React components, using a data-fetching pattern, or in API routes, or in `getStaticProps`, `getServerSideProps`, even `getInitialProps`, but that's not recommended these days.

Again, the custom hook, is only for React function components. `fetchData` is the star of the show.

> 2. PageC will be a regular 'react' component without any getStaticProps or getServerSideProps. Correct?

Yes. And because (assuming) `_app`, doesn't implement `getInitialProps`, Next.js automatically optimises the page by generating HTML for it, at build time. However, this page has JavaScript (the code you write), to fetch data from the client, and update the data it shows, shortly after the user visits the page, but the page initial HTML remains unchanged.

> 3. When we have pagination, it will introduce another variable parameter like 'pageNumber' to the API.  We can determine value of this only based on user actions on UI.  So, in this case, is it still possible to have getStaticProps on PageA and PageB?

There's some implications.

`getStaticProps` runs at build time, and for example, could be run using `pageNumber=0`, so then the initial HTML of page A and page B, will be based on what `pageNumber=0` returned at the time the pages are built. 

Even if the client goes and fetches a most recent `pageNumber=0` at the time the user visits the page. **This might be a problem for SEO.** 

Then from the client, if you user wants to see `pageNumber=2`, you'd have `fetchData`, make a request to get that page and update what the user sees. The initial HTML will still be the `pageNumber=0` at the time it was built.

There's ISR, which allows you to re-run getStaticProps at runtime, and regenerate what users see at `pageNumber=0`, https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration. So SEO sees updates happening to your pages A and B, without having to re-deploy the application.

**Reply by vasanthi-d** [reply]
Thank you again for your detailed response. I'll try it out and post the result here.

**Reply by icyJoseph** [reply]
One thing I realize was wrong is question 2. Pages with getStaticProps do get optimistic even if getInitialProps exists in _app.
